Various micro-optimizations in the URL code

MaxKellermann opened the pull request.

fritsch commented:
jenkins build this please

MrMC commented:
you could make the in-lines all on one line to match code convention used in other classes. Otherwise, the code convention is "{" and "}" on separate lines :)

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -4043,9 +4033,9 @@ bool CMusicDatabase::GetAlbumsByWhere(const std::string &baseDir, const Filter &
+    for (const auto &i : results)
     {
-      unsigned int targetRow = (unsigned int)it->at(FieldRow).asInteger();
+      unsigned int targetRow = (unsigned int)i.at(FieldRow).asInteger();
```
Would you mind also updating lines you've touched to use c++ style casts? Can be painful with several commits so this is me asking nicely, not demanding :)

Paxxi commented:
Nice work @MaxKellerman, always nice getting some modernization done. The inlining of methods will probably be a noop, compiler should figure this out better than us. It doesn't hurt either, but if it's to be merged it needs to follow coding style as @davilla mentioned

MaxKellermann commented:
On 2016/07/04 21:46, Pär Björklund <notifications@github.com> wrote:
> Nice work @MaxKellerman, always nice getting some modernization
> done. The inlining of methods will probably be a noop, compiler
> should figure this out better than us.

No, that's impossible.  The compiler can't optimize across compilation
units.

(That is, unless you use LTO, which is a workaround for the
"compilation unit" thing.)

MaxKellermann commented:
Second round:
- fixed the `TestBasicEnvironment` build failure, which was not caused, only triggered by my include cleanups;
- dropped the `std::to_string()` patch (for now), because `std::to_string()` is broken on the old NDK your Jenkins server is using

MaxKellermann committed:
MusicDatabase: use range-based "for"

MaxKellermann committed:
MusicDatabase: use "auto"

MaxKellermann committed:
VideoDatabase: use range-based "for"

MaxKellermann committed:
VideoDatabase: use "auto"

MaxKellermann committed:
DbUrl: use "override"

MaxKellermann committed:
DbUrl: make the base class "private"

Nobody ever casts to the base class.  By making it "private", we
narrow the interface, reducing it to a few explicitly mentioned APIs.
This will allow further optimizations, such as making more methods
non-virtual.

MaxKellermann committed:
utils/UrlOptions: remove explicit std::string initialization to empty

This is implied, and it's cheaper.

MaxKellermann committed:
URL: inline trivial methods

These are so simple that a function call is bigger than its inline.

Not inlining the destructor because there may be a bloat of implicitly
generated code in there.

MaxKellermann committed:
utils/UrlOptions: make several methods non-virtual

There are no overrides.

MaxKellermann committed:
utils/UrlOptions: use range-based "for"

MaxKellermann committed:
utils/UrlOptions: include cleanup

MaxKellermann committed:
utils/UrlOptions: use std::string::clear()

This is cheaper than assigning an empty string.

MaxKellermann committed:
utils/Variant: include cleanup

MaxKellermann committed:
utils/Variant: use std::vector::emplace_back()

MaxKellermann committed:
DbUrl: include cleanup

MaxKellermann committed:
utils/UrlOptions: use "auto"

MaxKellermann committed:
test/TestBasicEnvironment: include cores/DataCacheCore.h

Needed because gcc 4.8 complains about the use of std::unique_ptr<>
with a forward-declared type:

 /home/travis/build/xbmc/xbmc/xbmc/ServiceManager.h:47:7:   required from here
 /usr/include/c++/4.8/bits/unique_ptr.h:65:22: error: invalid application of ‘sizeof’ to incomplete type ‘CDataCacheCore’

Newer gcc versions do not have this problem, but unfortunately, the
Kodi build server runs this old version.

MaxKellermann commented:
Third round: another attempt for the `TestBasicEnvironment` build failure, which turns out to be a gcc 4.8 bug (I don't have such an old buggy version, but your build server does). Workaround: add an include dependency which should not be necessary ...

MartijnKaijser commented:
jenkins build this please

MartijnKaijser commented:
Any objections?

MartijnKaijser commented:
@MaxKellermann here's a run with test enabled http://jenkins.kodi.tv/job/TestMulti-All/668/

Paxxi commented:
👍
